File: servers/nextjs/app/(presentation-generator)/services/api/community.ts
import { ApiResponseHandler } from "./api-error-handler";
import { getHeader } from "./header";
import { getApiUrl } from "@/utils/api";

export interface CommunityPresentation {
  id: number;
  title?: string | null;
  description?: string | null;
  created_by?: string | null;
  likes?: number | null;
  views?: number | null;
  slides?: string[];
  fonts?: Record<string, string> | null;
  prompt?: string | null;
  v2_presentation?: string | number | null;
  community_reference_ids?: Array<string | number> | null;
  setup?: CommunityPresentationSetup | null;
}

export interface CommunityPresentationSetup {
  text_provider?: string | null;
  text_model?: string | null;
  image_provider?: string | null;
  web_search_provider?: string | null;
}

export interface CommunityPresentationListResponse {
  total_pages: number;
  page: number;
  page_size: number;
  results: CommunityPresentation[];
}

export type CommunityPresentationOrderBy =
  | "created_at"
  | "views"
  | "likes"
  | "priority";

export type CommunityPresentationSortOrder = "asc" | "desc";

export interface CommunityPresentationListFilters {
  created_at_gt?: string;
  created_at_lt?: string;
  views?: number;
  views_gt?: number;
  views_lt?: number;
  likes?: number;
  likes_gt?: number;
  likes_lt?: number;
  order_by?: CommunityPresentationOrderBy;
  order?: CommunityPresentationSortOrder;
}

export class CommunityPresentationApi {
  static async list(
    page = 1,
    pageSize = 8,
    signal?: AbortSignal,
    filters: CommunityPresentationListFilters = {}
  ): Promise<CommunityPresentationListResponse> {
    const params = new URLSearchParams({
      page: String(page),
      page_size: String(pageSize),
      order_by: filters.order_by ?? "priority",
      order: filters.order ?? "desc",
    });

    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== "") {
        params.set(key, String(value));
      }
    });

    const response = await fetch(
      getApiUrl(`/api/v1/ppt/community/presentations?${params.toString()}`),
      {
        headers: getHeader(),
        cache: "no-cache",
        signal,
      }
    );
    return ApiResponseHandler.handleResponse(
      response,
      "Failed to load community references"
    );
  }

  static async getById(id: number): Promise<CommunityPresentation> {
    const response = await fetch(
      getApiUrl(`/api/v1/ppt/community/presentations/${id}`),
      { headers: getHeader(), cache: "no-cache" }
    );
    return ApiResponseHandler.handleResponse(
      response,
      "Failed to load the community reference"
    );
  }
}

export function getCommunityPresentationTitle(
  presentation: CommunityPresentation
) {
  return presentation.title?.trim() || "Untitled presentation";
}

export function getCommunityPresentationAuthor(
  presentation: CommunityPresentation
) {
  return presentation.created_by?.trim() || "Presenton";
}

export function getCommunityReferenceIds(
  presentation: CommunityPresentation | null
) {
  if (!presentation?.community_reference_ids) return [];

  return Array.from(
    new Set(
      presentation.community_reference_ids
        .map((id) => Number(id))
        .filter((id) => Number.isSafeInteger(id) && id > 0)
    )
  );
}

File: servers/fastapi/api/v1/ppt/endpoints/community.py
from typing import Any, Literal

from fastapi import APIRouter, Path, Query

from services.community_presentations import (
    get_community_presentation,
    list_community_presentations,
)


COMMUNITY_ROUTER = APIRouter(prefix="/community/presentations", tags=["Community"])


@COMMUNITY_ROUTER.get("")
async def list_presentations(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=8, ge=1, le=24),
    created_at_gt: str | None = None,
    created_at_lt: str | None = None,
    views: int | None = Query(default=None, ge=0),
    views_gt: int | None = Query(default=None, ge=0),
    views_lt: int | None = Query(default=None, ge=0),
    likes: int | None = Query(default=None, ge=0),
    likes_gt: int | None = Query(default=None, ge=0),
    likes_lt: int | None = Query(default=None, ge=0),
    order_by: Literal["created_at", "views", "likes", "priority"] = "priority",
    order: Literal["asc", "desc"] = "desc",
) -> dict[str, Any]:
    return await list_community_presentations(
        page=page,
        page_size=page_size,
        created_at_gt=created_at_gt,
        created_at_lt=created_at_lt,
        views=views,
        views_gt=views_gt,
        views_lt=views_lt,
        likes=likes,
        likes_gt=likes_gt,
        likes_lt=likes_lt,
        order_by=order_by,
        order=order,
    )


@COMMUNITY_ROUTER.get("/{community_id}")
async def get_presentation(
    community_id: int = Path(ge=1),
) -> dict[str, Any]:
    return await get_community_presentation(community_id)
